### ISSUE TITLE
Regression in descriptor complex decorators

### ISSUE THREAD

**Issue by Rapptz** [microsoft/pyright#1903]
**Describe the bug**

When a complex decorator returns a descriptor the type checker no longer realises the attribute exists.

**To Reproduce**

Make a complex decorator that returns a descriptor and apply to a member function.

**Expected behavior**

The attribute exists.

**Screenshots or Code**

```python
from __future__ import annotations
from typing import Any, Callable, Generic, Optional, Type, TypeVar, overload

T = TypeVar('T')
T_co = TypeVar('T_co', covariant=True)

class Minimal(Generic[T, T_co]):
    def __init__(self, name: str, func: Callable[[T], T_co]):
        ...

    @overload
    def __get__(self, instance: None, owner: Type[T]) -> Minimal[T, T_co]:
        ...

    @overload
    def __get__(self, instance: T, owner: Type[T]) -> T_co:
        ...

    def __get__(self, instance: Optional[T], owner: Type[T]) -> Any:
        ...

def minimal_property(name: str) -> Callable[[Callable[[T], T_co]], Minimal[T, T_co]]:
    def decorator(func: Callable[[T], T_co]) -> Minimal[T, T_co]:
        return Minimal(name, func)

    return decorator

class B:
    @minimal_property('foo')
    def foo(self) -> str:
        return 'hello'

b = B()
b.foo # error
```

**VS Code extension or command-line**

Pylance 2021.5.3

**Additional context**

This code used to work in 2021.5.2. I'm unsure what changed. Note that without a complex decorator the code passes type checks just fine: 

![image](https://user-images.githubusercontent.com/1695103/119284265-48635c00-bc0d-11eb-9beb-751c3688d0d0.png)

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. We have pretty good test coverage for descriptors, but this particular case exposed a hole in our unit tests. The problem wasn't with the descriptor but rather with your use of the `Callable[[Callable[[T], T_co]], Minimal[T, T_co]]` return type annotation for the `minimal_property` decorator. This is an interesting case to handle. PEP 484 indicates that type variables must have a defined scope — a function, class or type alias. A `Callable` type is technically none of these, yet it is defining the scope for type variables `T` and `T_co` in this case, and we need to solve for these type variables when the decorator is applied. Pyright contains special logic to handle this situation, but it relies on certain assumptions, and I broke one of those assumptions with a recent bug fix.

In any case, the problem is now fixed, and I've added your code sample above to the unit tests so a similar regression is prevented in the future.
